### DISCUSSION TITLE
New image manipulation option (crop_scale)

### DISCUSSION THREAD

**Discussion by MatthewJLewis** [ExpressionEngine/ExpressionEngine#2924] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
In responsive, mobile-first design, it's often desirable to have images that are the same proportion to one another. In many "listing" types of pages, responsive layouts can end up looking very wonky on certain displays if the entry images are not consistently the same width and height.

**Describe the solution you'd like**
A solution I'd like to see would be a new image manipulation in ExpressionEngine that would allow us to set a width and a height, then it would center the existing image on a canvas of that size. (The centering could also be a configurable option; for instance, some implementations might prefer the image to go at the top left or center right of the canvas.)

**Describe alternatives you've considered**
The only real alternatives to this are to do the necessary manipulations to images prior to uploading to EE, which is cumbersome and prone to error, or else forgo using some very nice design options.

**Comment by jcogs-design** [comment]
This is a good idea.  I've been kicking around a mod to the built in image handling for EE that would add a 'crop_scale' option to the 'crop' and 'constrain' choices offered.  Also thought about enabling other image formats provided by GD2 at same time.  Would appear to require only some small changes to do.  Not got time right now, but will pick it up at some point if none else do.

In the interim, you can achieve this effect using CE-Image - which is my work-around.

**Comment by benoitlemay** [comment]
+1

**Comment by ilabcrossmedia** [comment]
+1

**Comment by aasun** [comment]
+1

**Comment by jcogs-design** [comment]
I've had a quick look to see how this **crop_scale** feature might be coded.

Image effects in EE6 appear to be handled by the [File controller](https://github.com/ExpressionEngine/ExpressionEngine/blob/6.dev/system/ee/ExpressionEngine/Controller/Files/File.php) which in turn calls a legacy library [filemanager](https://github.com/ExpressionEngine/ExpressionEngine/blob/36581fd1243c5845c125056bf8cf63406cd4d9d4/system/ee/legacy/libraries/Filemanager.php) to handle the actual image manipulations ([line 189](https://github.com/ExpressionEngine/ExpressionEngine/blob/36581fd1243c5845c125056bf8cf63406cd4d9d4/system/ee/ExpressionEngine/Controller/Files/File.php#L189)).  `Filemanager` in turn uses another legacy library ([image_lib](https://github.com/ExpressionEngine/ExpressionEngine/blob/36581fd1243c5845c125056bf8cf63406cd4d9d4/system/ee/legacy/libraries/Image_lib.php)) to do the actual resizing / cropping etc.  Looks like these libraries have been updated to also handle dynamic image manipulations, so if these libraries are updated then **crop_scale** could be come available to both static and dynamic file manipulations.

This suggests that what would appear to be required is:

- Modify the `image_process_*` functions in `Image_lib` to add support for the **crop_scale** option - this in turn looks like simply adding an extra branch to the conditional with code (e.g. `elseif ($action == 'crop_scale') {...}`) to set the appropriate height / width / x-origin / y-origin values for the image concerned for each of GD2, Imagemagick and netpbm variations of the function; further it would appear that much of the code required to determine the appropriate crop_scale transformation values for height / width / x-origin / y-origin could be borrowed from the function [max_hw_check](https://github.com/ExpressionEngine/ExpressionEngine/blob/36581fd1243c5845c125056bf8cf63406cd4d9d4/system/ee/legacy/libraries/Filemanager.php#L499) in the Filemanager library.
- Modify the function [edit_image](https://github.com/ExpressionEngine/ExpressionEngine/blob/36581fd1243c5845c125056bf8cf63406cd4d9d4/system/ee/legacy/libraries/Filemanager.php#L2187) in the library `Filemanager` to reflect the addition of a `crop_scale` option.
- Modify the function [crop](https://github.com/ExpressionEngine/ExpressionEngine/blob/36581fd1243c5845c125056bf8cf63406cd4d9d4/system/ee/ExpressionEngine/Controller/Files/File.php#L96) in the Controller [File](https://github.com/ExpressionEngine/ExpressionEngine/blob/36581fd1243c5845c125056bf8cf63406cd4d9d4/system/ee/ExpressionEngine/Controller/Files/File.php) to include a **crop_scale** option

I have not had time to research where these changes would need to be made (but I'm guessing [View/Files](https://github.com/ExpressionEngine/ExpressionEngine/tree/6.dev/system/ee/ExpressionEngine/View/files) might be a good place to start)
- Some modifications to the CP files interface to either add a **crop_scale** button to the file listing or to embellish the existing **crop** screen to include *crop_scale*
- Some modifications to the CP upload interface to add **crop_scale** as an option
- Some changes to the relevant sections of documentation

It's not yet the whole story: it really needs some input from someone familiar with how the CP is coded (in the absence of documentation), but hopefully will provide a useful starting point should anyone have the time / inclination to actually code this up.

**Comment by robsonsobral** [comment]
There's a [JS lib](https://github.com/jwagner/smartcrop.js) which have been [ported to PHP](https://github.com/xymak/smartcrop.php) and could be a good default option to center the crop.

![image](https://user-images.githubusercontent.com/931642/120027274-8b388180-bfc9-11eb-99c2-1654daf46571.png)

**Comment by robsonsobral** [comment]
It looks like the parameter `maintain_ratio` of the on-the-fly [:resize](https://docs.expressionengine.com/latest/fieldtypes/file.html#resize) and the [:resize_crop](https://docs.expressionengine.com/latest/fieldtypes/file.html#resize_crop) do a similar thing.

The function [`image_reproportion()`](https://github.com/ExpressionEngine/ExpressionEngine/blob/36581fd1243c5845c125056bf8cf63406cd4d9d4/system/ee/legacy/libraries/Image_lib.php#L1313) can be used.

**Comment by pyrobob83** [comment]
@intoeetive @TomJaeger  Bumping this discussion after testing #2947 and still not being able to do what I hoped.  If I'm understanding the original request here correctly, the desire is the same as what I want: to replicate traditional upload directory manipulations, which resize_crop seemingly cannot do effectively without introducing black bars for certain aspect ratios.

With a simple File Grid field containing 4 images with different aspect ratios (letterbox, landscape, portrait and square), consider this code:

```html
<p class="mb-4 text-center">Original source images - shows original apsect ratio</p>
<div class="mb-8 flex flex-wrap justify-center gap-5">
    {file_grid}
        <div class="basis-60 flex flex-col items-center gap-2 border border-gray2">
            <p>{file_grid:alt}:</p>
            <div class="aspect-square w-full relative">
                <img class="absolute w-full h-full object-contain" src="{file_grid:file}" />
            </div>
        </div>
    {/file_grid}
</div>

<p class="mb-4 text-center">Desired output - this uses traditional manipulations with type=crop, width=240 and heigh=160</p>
<div class="mb-8 flex flex-wrap justify-center gap-5">
    {file_grid}
        <div class="basis-60">
            <img src="{file_grid:file:thumbnail}" />
        </div>
    {/file_grid}
</div>

<p class="mb-2 text-center">Closest I can get using resize_crop - uses these params:</p>
<p class="mb-4 text-center">{ file_grid:file:resize_crop resize:width='240' crop:width='240' crop:height='160' position='center' crop:maintain_ratio='n' }</p>
<div class="mb-8 flex flex-wrap justify-center gap-5">
    {file_grid}
        <div class="basis-60">
            <img src="{file_grid:file:resize_crop resize:width='240' crop:width='240' crop:height='160' position='center' crop:maintain_ratio='n'}" />
        </div>
    {/file_grid}
</div>
```

which gives this output:
<img width="533" alt="image" src="https://github.com/ExpressionEngine/ExpressionEngine/assets/8193519/7416696a-8252-4c22-a559-6a71c3565f28">

Can we please reconsider this new on-the-fly manipulation type to replicate what traditional manipulations can do?  Or add parameters to resize_crop that can get the same result?

**Reply by pyrobob83** [reply]
@TomJaeger @intoeetive Bump!  Running into needing this once again and there's no way to replicate how traditional manipulations work with on-the-fly ones.

**Reply by intoeetive** [reply]
@pyrobob83 sorry that it took me so long, but there IS a way to achieve what you need. TBH, took me a while to discover, so I will update the docs with an example

TL; DR: `{file_grid:file:resize_crop resize:height='160' crop:width='240' crop:height='160' position='center' crop:maintain_ratio='n' crop:master_dim="height"}`

First, resize the image to desired height. Then, crop it to dimensions that you need, BUT you need to explicitly tell the system that keeping the height is the priority. Thus, we add  `master_dim="height"` parameter - and voila!

**Reply by intoeetive** [reply]
@pyrobob83 if the source image proportions can be very different (so master_dim cannot be set) here's an update. It requires both height and width to be set, e.g. `{file_grid:file:resize_crop width='240' height='160' position='center' crop:maintain_ratio='n'}`

https://github.com/ExpressionEngine/ExpressionEngine/pull/4988

**Comment by pyrobob83** [comment]
@intoeetive Thanks for looking at this and for the suggestions, but neither of these work the same as the standard manipulations:

#2 doesn't work because it adds black bars to fill the space (which certainly doesn't happen with standard manipulations) - here's an example that has different aspect ratios: `{gallery:file:resize_crop width='350' height='350' position='center' crop:maintain_ratio='n'}` produces this:
<img width="372" height="373" alt="image" src="https://github.com/user-attachments/assets/8a442310-5284-422c-8043-e423c1a75250" />


And #1 doesn't work because there _are_ varying aspect ratios in the set, so there isn't a consistent axis to resize against.  When you run into a scenario that doesn't match what you set, you get black bars again.  Same example, but with `{gallery:file:resize_crop resize:height='350' crop:width='350' crop:height='350' position='center' crop:maintain_ratio='n' crop:master_dim='width'}` gives this:
<img width="379" height="376" alt="image" src="https://github.com/user-attachments/assets/c3f54b6f-41a7-4feb-b449-75fbd879729c" />

**Reply by pyrobob83** [reply]
Programatically, I think what we need is a way to first crop to a specific **_aspect ratio_** (ie the desired final size), _**then**_ do the resize.  Does that make sense?

**Reply by intoeetive** [reply]
@pyrobob83 I re-tested the branch https://github.com/ExpressionEngine/ExpressionEngine/pull/4988 and it works well for me with different aspect ratios. Can you please try again, make sure you delete the existing manipulation files (`_resize` and `_crop` folders in your upload location) and also force-refresh the browser

If you want to crop then resize, you can always do that by chaining `:crop:resize` but that wouldn't work well for most of the cases

**Comment by pyrobob83** [comment]
@intoeetive Bah, sorry, my bad - I missed you were referencing a specific branch so was just testing the latest public release.  Will test #4988  as soon as I can.

**Comment by pyrobob83** [comment]
@intoeetive Ok, your branch #4988 does indeed do exactly what I was hoping with `{gallery:file:resize_crop width='350' height='350' position='center' crop:maintain_ratio='n'}`.  Thank you!!!

**Comment by pyrobob83** [comment]
@intoeetive After further testing, the efficiency the page doesn't seem right - as though the caching isn't working.  With my old setup using `{gallery:file:resize height='350' width='10000' quality='90'}` (ie just resizing to 350px wide for all images), I get about a 7-8 second load time after clearing the `_resize` dir, then consistently around 0.8s for subsequent loads... but with the patch from #4988 and using `{gallery:file:resize_crop width='350' height='350' position='center' crop:maintain_ratio='n' quality='80'}`, I seem to get that 7-8s load time every time.  The images don't appear to be regenerated on subsequent loads (based on the modified date not changing), so I'm not sure what this is... any ideas?
